README.md
# Exile

Created by Dokimi

For Minetest 5.4

### Installation
Installation from the ContentDB is preferred, just add it from the in-game menu.

If extracting manually, Exile should be in games folder like so:

Minetest > games > exile > mods/menu/... etc

Since 0.2.3, Exile requires naturalslopeslib. If you're installing the source
from git, you'll need to install nsl as well. Extract it into the exile mods/
folder or use 'git pull --recurse submodules' if you have cloned the repository.

### Settings for multiplayer
"time_speed" defaults to 72, at this rate a player who logs on at 8:00am every
day will get a change in season every 2.5 days. Changing speed to 60 will make
days last 24 minutes and a new season every day, but he will see the seasons
in reverse. At time_speed of 96, days last only 15 minutes, but the player
might see spring (year 1) on day 1, summer (year 2) on day 2, etc.

Set exile_hud_update to 1 second for multiplayer servers on the internet; a
lan server can probably handle 0.2 seconds.

### Mods for multiplayer
Wield3d is recommended: https://github.com/stujones11/wield3d
Alternode was used to add the "infotext" popups in the spawn shelter on the
Land of Catastrophe server: https://github.com/AntumMT/mod-alternode


### Gameplay
Challenging, at times brutal, wilderness survival with simple technology.
Use your wits to find food, water, and shelter before succumbing to the elements,
while exploring the mysterious world, and developing your capacities to endure your exile.

Features:
Player health effects: hypothermia, exhaustion, disease...
Dynamic nature: Seasonal weather, erosion, water flows through soil...
Building matters: make shelters from the rain, kilns, smelters,...

CAUTION: some people may find parts of this game difficult or disturbing.

### Gameplay Guide
Check out doc > walkthrough for a more detailed guide.

Many different strategies might work, and part of the fun is figuring out what does,
and catastrophically does not, work.

Here are some early steps you might pass though:
1. Make basic tools. Find a suitable camp site soon before you get tired.
2. Build a bed to rest in, shelter, and maybe fire for warmth.
3. Make a kiln, and make pots. Harvest wild foods while you wait.
4. Collect water with pots, and build up a water supply.
5. Farm food, drugs, and fibre to build up supplies.
6. Explore and gather resources for more advanced tools.

You start in spring. Soon the hot dry weather of summer will arrive.
Running out of water, or getting exhausted by the heat is a real risk.
After that will come the sub-zero conditions of winter.
Starvation and freezing are hard to avoid without preparation.


Some tips and tricks:
- Beds are important. If you're exhausted you get hypothermia/heat stroke. Get under shelter in a nice temperature.
- Do your crafting, organizing, planning while you rest.
- Weather: extremes sap your energy. Like real life, walking around in a snowstorm is a bad idea!
- Temperature: Build a shelter, with a fire place or lots of torches. You can also go underground, caves are safe from weather.
- Water: you can drink cave drips (click them). Water pots collect rain water. Some plants, and food quench your thirst. If you're desperate, you can try to melt ice, or dig a seepage pit in wet ground and wait for water to flow into it
- Food: eat stuff and see if you live! You can catch animals with clubs (right click)
- Clothing: Simple grass clothing can be woven quickly, and can be a lifesaver. Better clothes can be made by soaking bundles of the right kind of plants.
- Farming: digging tools also can till soil. Punching depleted farm soil with fertilizers will restore it.
- Keep an eye on "health effects". You may have eaten something bad, or have a terrible disease.
- Drugs: some plants have useful medicinal effects. Be careful not to overdose, however.
- Go deep enough underground and you might find something....
- Hardwood trees are more difficult to cut than soft wood, you will need better tools.
- Build stairs and shelters around your base, this will save you energy and protect you from extremes.
- You can use sticks to build a ladder, or a pole to shimmy up, to get to high places, or descend if you're careful.
- Not every step in crafting can be done at a work station. Some things need to be fired, or soaked in water etc.
- Ovens, kilns, furnaces: build them like you would for real. A fire with access to air, and a sealed chamber that gets heated up.
- Fires. Blocks are hotter. Charcoal is hotter than wood. Fires can be temporarily extinguished by punching them holding sediment.
- Charcoal. Make it like you would for real: a wood fire sealed up with no air.
- Glassmaking: Sand and wood ash can be made into green glass; the ash must be soaked, dried, and roasted to make clear glass. Glass can be melted onto iron trays to make panes for real windows!
- Iron smelting. This is hard. It needs plenty of charcoal, and a space below the iron mixture for slag to drain out.


### Development
Exile is opensource - that means it is as good as you choose to make it.
It also means development can be erratic and haphazard at times, so be patient!

Exile is currently in Alpha, therefore expect there can be bugs, missing features, performance issues,
and perhaps compatibility breaking updates.
Despite this, Exile does have enough features to be a playable game and should be stable and mostly
bug-free.

See the git repository at
https://github.com/jeremyshannon/Exile/ for known bugs, and to report new ones.

### Credit
Thanks is due to all those whose mods have been adapted for use in Exile (see mod folders for details).

Thanks also to all who have given feedback, fixes etc. A full, up-to-date list of contributors can be
found on the github repository, under the Insights tab.




CHANGELOG.md
# Changelog:

## [2022-01-21][2022-01-21__0.3.5] — 0.3.5 and 0.2.8
  - Woven blankets to keep exiles warmer early in the game.
  - Foods can now burn if left cooking too long.
  - Mobs should no longer `stun`-lock when hit repeatedly.
  - Nerf: Increased fall damage.
  - You can use a `Stick` to bar a door, to prevent entry.
  - New animations, thanks to ***MisterE***.
  - Track how many incarnations you’ve had in the `Character` tab.
  - `Glow Worm`s in caves.
  - Nerf: `Stick`s don’t connect with right-angle corners anymore.
  - Nerf: `Mudbrick` becomes clay when broken, in the same manner as bricks lose their mortar.
  - Mapgen: `carpathian` is now supported, for a harder game world.
  - No more beach spiders.
  - New food-system infrastructure.
  - Experimental biomes — enable under global settings to test them.
  - Miscellaneous bug-fixes and quality of life improvements.
   
## [2021-11-01][2021-11-01__0.3.4] — 0.3.4 and 0.2.7
  - New player API: Head tracking, multiplayer skins, male and female characters.
  - Added an introduction screen for new players, and a chat display of the day/season/year with in-universe style.
  - Place snow/ice/water on top of a clay `Water Pot`, and melting will fill it.
  - Buff: Increased food-value of meat.
  - Improved `Torch` stacking by rounding-off burn times.
  - New chat commands: `/set_tempscale`, `/date`.
  - Better indoors-check on `Minetest 5.4.x`.

## [2021-09-25][2021-09-25__0.3.3] — 0.3.3 and 0.2.6
  - Plants now grow even when players are away from the area.
  - Fireproof: `Iron Doors` and `Iron Trapdoors`.
  - Ability to mix `Wet Loam` and agricultural soil.
  - Renamed `Stone Chopper` to `Stone Knife` for clarity.
  - More bug-fixes and improvements.

## [2021-09-12][2021-09-12__0.3.2] — 0.3.2 and 0.2.5
  - Glassworking! Make windows, and make glass bottles that can hold water, and which stack.
  - Agricultural soils erode down to a soil slab instead of vanishing.
  - Fixed a crash related to baking foods.

## [2021-07-18][2021-07-18__0.3.1] — 0.3.1 and 0.2.4
  - `Oil Lamp`s can be switched off with right-click, and lit with `Fire Stick`s.
  - `Oil Lamp`s can hold more oil and now burn oil slower.
  - Various bug-fixes.

## [2021-06-12][2021-06-12__0.3.0] — 0.3.0 and 0.2.3
  - Initial release of `Minetest 5.4.0`; use Exile 0.2 branch for the `5.3.0` series.
  - New dependency: `naturalslopeslib` provides better hillsides.
  - Ability to throw `Torch`es for light.
  - Settings for `breaktaker`.
  - Ability to display temperature in Celsius, Fahrenheit, or Kelvin. (Using global settings.)
  - Food burns instead of vanishing, thanks to ***ts***.
  - Farming improvements.
  - Fixes for fire and rain.
  - Many fixes for crashes and bugs.

## [2021-02-15][2021-02-15_Mantar] — Now maintained by Mantar

## [2020-12-29][2020-12-29__0.2.2] — 0.2.2
  - New HUD.
  - Character tab.
  - Health effects.
  - Drugs: `Tiku`, `Tang`, `Metastim`.
  - Surface air-shafts via `megamorph`.
  - Fixed clothing infinite tolerance bug.
  - Added `Iron Fittings` for carpentry products.
  - Young vs. old `Tangkal` trees.
  - Added `Tiken` woody cane.
  - Cave sediments and mushrooms.
  - Defined `puts_out_fire` group (e.g. mud now puts out fire), and `masonry` group.
  - Created a walkthrough document for players who want one.
  - Added `Loose Wattle`.
  - Added `Mashed Anperla`.
  - Miscellaneous minor fixes and balancing.

## [2020-09-25][2020-09-25__0.2.1] — 0.2.1
  - Fixed exile-letter crash.
  - Fixed saltwater-surface sediment crash.
  - Fixed multiplayer health-tab crash.
  - Animals balanced and bug-fixed.
  - Adjusted ice freezing and thawing.

## [2020-08-02][2020-08-02__0.2.0] — 0.2.0
  - Clothing (plus multi-skin model, new player API), `Weaving Frame`, and temperature tolerance via clothing.
  - Lore: exile letter.
  - Added `Sleeping Spot`.
  - New mobs: `Darkasthaan`, `Impethu`, `Pegasun`, `Sneachan`.
  - Cookable animal carcasses.
  - Graffiti: glow paint.
  - New building materials: `Rammed Earth`, bricks and mortar, roof tiles, wooden doors/floors/stairs, mortared masonry.
  - Fixed dungeon loot.
  - Added more artifacts. (`Airboat`, art, transporter, `Wayfinder`, ….)
  - Added three artifact species of mushrooms for sustaining fibre, food/water, and sticks, absent sunlight.
  - Split Geomoria mod’s biomes into realms (highways, and city structures).
  - Added more `geomorphs`.
  - `Gneiss` and `Jade` for deep-underground biome.
  - Better-looking `Bones`.
  - Newly-formatted colouring for skies.
  - Chat commands for controlling climate variables.
  - `Design Guide`.
  - Updated menu-screen backgrounds.
  - Updated version of crafting mod.
  - Various minor fixes.
  - Various minor balance changes.
  - Various minor additions.
  - `Minetest 5.3.0` compatibility.

## [2020-03-11][2020-03-11__0.1.0] — 0.1.0
  - Initial release.

## [2020-03-10][2020-03-10_Dokimi] — Founded by Dokimi

[2022-01-21__0.3.5]: (https://github.com/DokimiCU/Exile/commits/5d3dfb4fc9e8636d57a00cb59df3b6c36edf4d9c)
 <!-- Unix time 1642873579 -->
 <!-- Fri Jan 21 16:46:19 2022 UTC-0800 -->
 <!-- 2022-01-22 00:46:19      UTC+0000 -->
 <!-- (https://content.minetest.net/packages/Mantar/exile/releases/10746/download/) -->

[2021-11-01__0.3.4]: (https://github.com/DokimiCU/Exile/commits/0889b4abd431fd578ea76a5fcc3a70aace7dfcb9)
 <!-- Unix time 1635744365 -->
 <!-- Sun Oct 31 18:26:05 2021 UTC-0700 -->
 <!-- 2021-11-01 01:26:05      UTC+0000 -->
 <!-- (https://content.minetest.net/packages/Mantar/exile/releases/9596/download/) -->

[2021-09-25__0.3.3]: (https://github.com/DokimiCU/Exile/commits/f5fc39c8b8ce7f6d2dec23caecbd6a59977d0520)
 <!-- Unix time 1632604522 -->
 <!-- Sat Sep 25 14:15:22 2021 UTC-0700 -->
 <!-- 2021-09-25 21:15:22      UTC+0000 -->
 <!-- This was the initial release of v0.3.3 but is now deprecated in favor of v0.3.3g -->
 <!-- (https://content.minetest.net/packages/Mantar/exile/releases/9461/download/) is v0.3.3g-->


mods/tech/drugs.lua
----DRUGS -- medicines etc
local random = math.random
----MEDICAL----
--Herbal medicine
-- removes energy cost of plants healing effects
--can heal certain health effects
--a restorative anti-bacterial/anti-parasitic
minetest.register_craftitem("tech:herbal_medicine", {
	description     = "Herbal Medicine",
	inventory_image = "tech_herbal_medicine.png",
	stack_max       = minimal.stack_max_medium *2,
	groups          = {flammable = 1},
	on_use          = function(itemstack, user, pointed_thing)
		local meta         = user:get_meta()
		local effects_list = meta:get_string("effects_list")
		effects_list       = minetest.deserialize(effects_list) or {}
		--remove parasites
		if random() < 0.33 then
			HEALTH.remove_new_effect(user, {"Intestinal Parasites"})
		end
		--cure/reduce food poisoning and infections
		--see how effective the dose is
		local cfp = random()
		if cfp < 0.25 then
			--cure up to severe
			HEALTH.remove_new_effect(user, {"Food Poisoning", 3})
			HEALTH.remove_new_effect(user, {"Fungal Infection", 3})
			HEALTH.remove_new_effect(user, {"Dust Fever", 3})
		elseif cfp < 0.5 then
			--cure up to moderate
			HEALTH.remove_new_effect(user, {"Food Poisoning", 2})
			HEALTH.remove_new_effect(user, {"Fungal Infection", 2})
			HEALTH.remove_new_effect(user, {"Dust Fever", 2})
		elseif cfp < 0.75 then
			--only cure mild
			HEALTH.remove_new_effect(user, {"Food Poisoning", 1})
			HEALTH.remove_new_effect(user, {"Fungal Infection", 1})
			HEALTH.remove_new_effect(user, {"Dust Fever", 1})
		end
		--hp_change, thirst_change, hunger_change, energy_change, temp_change, replace_with_item
		return HEALTH.use_item(itemstack, user, 5, 0, 0, 0, 0)
	end,
})
------------
--Detox?
--for toxins, alcohol, drugs
-- e.g. charcoal? - doesn't work for everything, can itself cause vomiting etc
--[[
	How toxins get treated:
	Alcohol: wait for it to pass (while helping them not die), pump stomach
	Stimulant OD: sedation, wait for it to pass (while helping them not die)
	Specific toxins can have anti-toxins (a fairly modern treatment)
]]
----DRUGS----
----STIMULANTS----
--Tiku -- stimulant drug -- gets you high
minetest.register_craftitem("tech:tiku", {
	description     = "Tiku (stimulant)",
	inventory_image = "tech_tiku.png",
	stack_max       = minimal.stack_max_medium *2,
	groups = {flammable = 1},
	on_use = function(itemstack, user, pointed_thing)
		--begin the bender
		HEALTH.add_new_effect(user, {"Tiku High", 1})
		--hp_change, thirst_change, hunger_change, energy_change, temp_change, replace_with_item
		return HEALTH.use_item(itemstack, user, 0, 0, -24, 96, 0)
	end,
})
----DEPRESSANTS----
--Tang, alcoholic drink
--Pot of Tang
minetest.register_node("tech:tang", {
description = "Tang",
	tiles = {
		"tech_pot_tang.png",
		"tech_pottery.png",
		"tech_pottery.png",
		"tech_pottery.png",
		"tech_pottery.png",
		"tech_pottery.png"
	},
	drawtype  = "nodebox",
	stack_max = 1, --minimal.stack_max_bulky,
	paramtype = "light",
	node_box  = {
		type  = "fixed",
		fixed = {
			{-0.2500,  0.3750, -0.2500, 0.2500,  0.5000, 0.2500}, -- NodeBox1
			{-0.3750, -0.2500, -0.3750, 0.3750,  0.3125, 0.3750}, -- NodeBox2
			{-0.3125, -0.3750, -0.3125, 0.3125, -0.2500, 0.3125}, -- NodeBox3
			{-0.2500, -0.5000, -0.2500, 0.2500, -0.3750, 0.2500}, -- NodeBox4
			{-0.3125,  0.3125, -0.3125, 0.3125,  0.3750, 0.3125}, -- NodeBox5
		}
	},
	groups = {dig_immediate = 3, pottery = 1, temp_pass = 1},
	sounds = nodes_nature.node_sound_stone_defaults(),
	on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
	--lets skull an entire vat of booze, what could possibly go wrong...
		local meta = clicker:get_meta()
		local thirst = meta:get_int("thirst")
		local hunger = meta:get_int("hunger")
		local energy = meta:get_int("energy")
		--only drink if thirsty
		if thirst < 100 then
			--you're skulling a whole bucket
			thirst = thirst + 100
			if thirst > 100 then
				thirst = 100
			end
			--all energy and half food equivalent of the fruit
			--gets given as energy
			energy = energy + 180
			if energy > 1000 then
				energy = 1000
			end
			hunger = hunger + 60
			if hunger > 1000 then
				hunger = 1000
			end
			--drunkness
			if random() < 0.75 then
				HEALTH.add_new_effect(clicker, {"Drunk", 1})
			end
			meta:set_int("thirst", thirst)
			meta:set_int("energy", energy)
			minetest.set_node(pos, {name = "tech:clay_water_pot"})
			minetest.sound_play("nodes_nature_slurp",
				{pos = pos, max_hear_distance = 3, gain = 0.25})
		end
	end
})
--save usage into inventory, to prevent infinite supply
local on_dig_tang = function(pos, node, digger)
	if minetest.is_protected(pos, digger:get_player_name()) then
		return false
	end
	local meta       = minetest.get_meta(pos)
	local ferment    = meta:get_int("ferment")
	local new_stack  = ItemStack("tech:tang_unfermented")
	local stack_meta = new_stack:get_meta()
	stack_meta:set_int("ferment", ferment)
	minetest.remove_node(pos)
	local player_inv = digger:get_inventory()
	if player_inv:room_for_item("main", new_stack) then
		player_inv:add_item("main", new_stack)
	else
		minetest.add_item(pos, new_stack)
	end
end
--set saved
local after_place_tang = function(pos, placer, itemstack, pointed_thing)
	local meta         = minetest.get_meta(pos)
	local stack_meta   = itemstack:get_meta()
	local ferment      = stack_meta:get_int("ferment")
	if ferment > 0 then
		meta:set_int("ferment", ferment)
	end
end
--Pot of new Tang, must be left to ferment
minetest.register_node("tech:tang_unfermented", {
description = "Tang (unfermented)",
	tiles = {
		"tech_pot_tang_uf.png",
		"tech_pottery.png",
		"tech_pottery.png",
		"tech_pottery.png",
		"tech_pottery.png",
		"tech_pottery.png"
	},
	drawtype  = "nodebox",
	stack_max = 1, --minimal.stack_max_bulky,
	paramtype = "light",
	node_box  = {
		type  = "fixed",
		fixed = {
			{-0.2500,  0.3750, -0.2500, 0.2500,  0.5000, 0.2500}, -- NodeBox1
			{-0.3750, -0.2500, -0.3750, 0.3750,  0.3125, 0.3750}, -- NodeBox2
			{-0.3125, -0.3750, -0.3125, 0.3125, -0.2500, 0.3125}, -- NodeBox3
			{-0.2500, -0.5000, -0.2500, 0.2500, -0.3750, 0.2500}, -- NodeBox4
			{-0.3125,  0.3125, -0.3125, 0.3125,  0.3750, 0.3125}, -- NodeBox5
		}
	},
	groups = {dig_immediate = 3, pottery = 1, temp_pass = 1},
	sounds = nodes_nature.node_sound_stone_defaults(),
	on_dig = function(pos, node, digger)
		on_dig_tang(pos, node, digger)
	end,
	on_construct = function(pos)
		--duration of ferment
		local meta = minetest.get_meta(pos)
		meta:set_int("ferment", math.random(300 , 360))
		--ferment
		minetest.get_node_timer(pos):start(5)
	end,
	after_place_node = function(pos, placer, itemstack, pointed_thing)
		after_place_tang(pos, placer, itemstack, pointed_thing)
	end,
	on_timer = function(pos, elapsed)
		local meta    = minetest.get_meta(pos)
		local ferment = meta:get_int("ferment")
		if ferment < 1 then
			minetest.set_node(pos, {name = "tech:tang"})
			--minetest.check_for_falling(pos)
			return false
		else
	--ferment if at right temp
	local temp = climate.get_point_temp(pos)
	if temp > 10 and temp < 34 then
		meta:set_int("ferment", ferment - 1)
	end
			return true
		end
	end,
})
---HALLUCINOGENS----
--Recipes -- mortar and pestle
--make herbal_medicine
crafting.register_recipe({
	type   = "mortar_and_pestle",
	output = "tech:herbal_medicine",
	items  = {"nodes_nature:hakimi", "nodes_nature:merki", "nodes_nature:moss"},
	level  = 1,
	always_known = true,
})
--make tiku
crafting.register_recipe({
	type   = "mortar_and_pestle",
	output = "tech:tiku",
	items  = {"nodes_nature:tikusati_seed 12", "nodes_nature:wiha", "tech:vegetable_oil"},
	level  = 1,
	always_known = true,
})
--make tang_unfermented
crafting.register_recipe({
	type   = "mortar_and_pestle",
	output = "tech:tang_unfermented",
	items  = {"nodes_nature:tangkal_fruit 12", "tech:clay_water_pot_freshwater"},
	level  = 1,
	always_known = true,
})
